packages/pwa-kit-react-sdk/src/ssr/universal/contexts/index.js

```diff
@@ -5,9 +5,57 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 
-import React from 'react'
+import React, {useEffect, useRef} from 'react'
+import PropTypes from 'prop-types'
+import {useLocation} from 'react-router-dom'
+import {uuidv4} from '../../../utils/uuidv4.client'
+
 import DeviceContext from '../device-context'
 
 const ExpressContext = React.createContext()
 
-export {DeviceContext, ExpressContext}
+const CorrelationIdContext = React.createContext()
+// regexp for uuid format
+const regexp = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i
+
+const CorrelationIdProvider = ({children, correlationId, resetOnPageChange = true}) => {
+    const _correlationIdFn = correlationId === 'function' && correlationId
+    const _correlationId = correlationId !== 'function' && correlationId
+    const [id, setId] = React.useState(_correlationId || _correlationIdFn())
+    const location = useLocation()
+
+    const isFirstRun = useRef(true)
+    useEffect(() => {
+        // this hook only runs on client-side
+        // don't run this on first render
+        if (isFirstRun.current) {
+            isFirstRun.current = false
+            return
+        }
+        if (resetOnPageChange && !_correlationIdFn) {
+            console.warn(
+                'correlationId needs to be a function returning a uuid string when resetOnPageChange is true '
+            )
+        }
```

Comment:
Does this need to be run every change of the effects source array? or can we been displayed once when the provider is used?